File: backend/topology.py
from dataclasses import dataclass, field
from typing import List
from enum import Enum


class NodeRole(str, Enum):
    ROUTER = "router"
    FIREWALL = "firewall"
    SERVER = "server"
    DB = "db"
    CLIENT = "client"
    SERVICE = "service"
    LB = "lb"
    CACHE = "cache"


class NodeTech(str, Enum):
    CISCO_IOS = "cisco_ios"
    NGINX = "nginx"
    HAPROXY = "haproxy"
    PYTHON = "python"
    NODEJS = "nodejs"
    MYSQL = "mysql"
    REDIS = "redis"
    LINUX = "linux"
    GENERIC = "generic"


@dataclass
class NetNode:
    id: str
    label: str
    role: NodeRole
    tech: NodeTech
    ip: str = ""
    os: str = ""
    x: float = 0
    y: float = 0


@dataclass
class NetEdge:
    id: str
    source: str
    target: str
    protocol: str = "TCP"
    port: int = 80
    critical: bool = True
    call_rate: float = 100.0


@dataclass
class SvcNode:
    id: str
    label: str
    tech: NodeTech
    role: NodeRole = NodeRole.SERVICE
    ip: str = ""
    x: float = 0
    y: float = 0


@dataclass
class SvcEdge:
    id: str
    source: str
    target: str
    protocol: str = "HTTP"
    port: int = 80
    critical: bool = True
    call_rate: float = 100.0


@dataclass
class Topology:
    net_nodes: List[NetNode] = field(default_factory=list)
    net_edges: List[NetEdge] = field(default_factory=list)
    svc_nodes: List[SvcNode] = field(default_factory=list)
    svc_edges: List[SvcEdge] = field(default_factory=list)

File: frontend/src/types/topology.ts
export type NodeRole =
  | "router"
  | "firewall"
  | "server"
  | "db"
  | "client"
  | "service"
  | "lb"
  | "cache";

export type NodeTech =
  | "cisco_ios"
  | "nginx"
  | "haproxy"
  | "python"
  | "nodejs"
  | "mysql"
  | "redis"
  | "linux"
  | "generic";

export interface TopologyNode {
  id: string;
  label: string;
  role: NodeRole;
  tech: NodeTech;
  ip: string;
  os: string;
  x: number;
  y: number;
}

export interface TopologyEdge {
  id: string;
  source: string;
  target: string;
  protocol: string;
  port: number;
  critical: boolean;
  callRate: number;
}

export interface TopologyData {
  netNodes: TopologyNode[];
  netEdges: TopologyEdge[];
  svcNodes: TopologyNode[];
  svcEdges: TopologyEdge[];
}

export type Selection = { type: "node" | "edge"; id: string } | null;
export type CanvasMode = "select" | "add-edge";

// Role → default tech
export const ROLE_DEFAULT_TECH: Record<NodeRole, NodeTech> = {
  router: "cisco_ios",
  firewall: "cisco_ios",
  lb: "haproxy",
  server: "nginx",
  service: "python",
  db: "mysql",
  cache: "redis",
  client: "generic",
};

// Node colour by role
export const NODE_COLOR: Record<NodeRole, string> = {
  router: "#3B82F6",
  firewall: "#3B82F6",
  lb: "#3B82F6",
  server: "#22C55E",
  service: "#22C55E",
  db: "#14B8A6",
  cache: "#14B8A6",
  client: "#F59E0B",
};

export const ELLIPSE_ROLES: NodeRole[] = ["db", "cache"];

// Toolbar node buttons
export const TOOLBAR_NODES: { label: string; role: NodeRole }[] = [
  { label: "Router / FW", role: "router" },
  { label: "Server",      role: "server"  },
  { label: "DB",          role: "db"      },
  { label: "Client",      role: "client"  },
  { label: "Service",     role: "service" },
  { label: "LB",          role: "lb"      },
  { label: "Cache",       role: "cache"   },
];
